File: 0216-combination-sum-iii/0216-combination-sum-iii.java
class Solution {
    Set<List<Integer>>ans;
    public List<List<Integer>> combinationSum3(int k, int n) {
        ans=new HashSet<>();
        solve(k,n,new ArrayList<>(),0,1);
        return new ArrayList<>(ans);
    }
    void solve(int k,int n,List<Integer> him,int sum,int ind){
        if(him.size()==k){
            if(sum==n){
                ans.add(new ArrayList<>(him));
                
            }
            return;
        }
        if(sum>n){
            return;
        }
        for(int i=ind;i<=9;i++){
            if(sum+i<=n){
                him.add(i);
                solve(k,n,him,sum+i,i+1);
                him.remove(him.size()-1);
                solve(k,n,him,sum,i+1);
            }
            
        }
    }
}

File: 0216-combination-sum-iii/0216-combination-sum-iii.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        self.a=[]
        self.solve(k,n,[],1)
        return self.a
    def solve(self,k,n,b,ind):
        if n==0 and k==0:
            self.a.append(b.copy())
            return
        if k==0:
            return
        for i in range(ind,10):
            if (n-i>=0):
                b.append(i)
                self.solve(k-1,n-i,b,i+1)
                b.pop()